### ISSUE TITLE
Application.SetVolume

### ISSUE THREAD

**Issue by jingai** [m0ngr31/kodi-alexa#19]
This is going to require a bit of thought.

We'd want to add volume up/down as well as the ability to set an absolute volume (0-100).  For volume up/down, we'll need to determine what is an acceptable granularity (10? 20?).

Additionally, we can support something like "set volume low/medium/high".

@digiltd has some existing code over at https://github.com/bwssytems/ha-bridge/issues/193 that might be of use here.

Note that this issue and the implementation of it are *only* to support adjusting *Kodi's* volume, not any connected A/V equipment.

**Comment by m0ngr31** [comment]
If we did low/medium/high, you could do 33% increments:
- 0%: mute
- 33%: low
- 66%: medium
- 100%: high

**Comment by Xaelias** [comment]
Why not just do it the Alexa way? 1 to 10?

**Comment by m0ngr31** [comment]
That works too.

**Comment by michaelhinchey** [comment]
Can this be done with harmony and ha bridge?

**Comment by jingai** [comment]
This particular Issue is to support adjusting volume in Kodi.  To control the volume of your A/V equipment you'd use whatever skill can do that; e.g., Logitech's Harmony skill (once it's released).

**Comment by digiltd** [comment]
@michaelhinchey This can be done separately from this skill. See https://github.com/bwssytems/ha-bridge/issues/193 for how I did it.

**Comment by jingai** [comment]
Kind of off-topic for this Issue, but I really wish "turn up/down" worked..

**Comment by digiltd** [comment]
@jingai You mean generally within Kodi or when using the ha-bridge?

**Comment by jingai** [comment]
I mean with home automation devices in general with Alexa.  At least, last time I tried, "turn up {device}" didn't brighten like I'd expect it to.  Does it now?

**Comment by michaelhinchey** [comment]
Yes but how would I incorporate that into the harmony hub?  I use my
receiver to adjust volume. Thx






         Every dream is a goal with a date on it; if you can conceive it
believe it, you can achieve it. Don't let anyone steal your dream. Make the
decision to take a positive attitude, take pride in yourself and what you
do, work hard and stay with it. A mind is a terrible thing to waste.


"I felt bad because I couldn't wheelie; until I met a man with no bicycle"

"You are young at any age... if you are planning on tomorrow"

**Comment by jingai** [comment]
So we don't derail this Issue, would you mind taking the discussion to the forum?  It's worth discussing, but this Issue is specifically about controlling Kodi's volume, not AV equipment.

**Comment by michaelhinchey** [comment]
I just tested alexa and she does recognize the  turn up and turn down command to adjust brightness and turn down fans.  I am still working on volume adjustment for my volume for harmony so that it works with Alexa.  Harmony skill doesn't allow this feature yet so I'm unsure how to get this to work with my harmony hub and harmony skill through alexa.   I see that people can adjust it through kodi remote but I use my receiver to adjust volume.   I know it's probably less complicated but I can't find any answers yet

**Comment by jingai** [comment]
Yes, there are likely a bunch of different ways to do it, but it isn't related to this issue.

**Comment by michaelhinchey** [comment]
Oh geez. I thought this was the forum.  What is the actual site.  Sorry






         Every dream is a goal with a date on it; if you can conceive it
believe it, you can achieve it. Don't let anyone steal your dream. Make the
decision to take a positive attitude, take pride in yourself and what you
do, work hard and stay with it. A mind is a terrible thing to waste.


"I felt bad because I couldn't wheelie; until I met a man with no bicycle"

"You are young at any age... if you are planning on tomorrow"

**Comment by m0ngr31** [comment]
http://forum.kodi.tv/showthread.php?tid=254502

**Comment by michaelhinchey** [comment]
Ah I see.  That is more or less a kodi forum.  I got mixed up with another
forum.  I need to be on either a harmony or amazon echo or ha bridge forum.






         Every dream is a goal with a date on it; if you can conceive it
believe it, you can achieve it. Don't let anyone steal your dream. Make the
decision to take a positive attitude, take pride in yourself and what you
do, work hard and stay with it. A mind is a terrible thing to waste.


"I felt bad because I couldn't wheelie; until I met a man with no bicycle"

"You are young at any age... if you are planning on tomorrow"

**Comment by jingai** [comment]
So, if I were to implement this in the same way Amazon does, the list of utterances would be:

turn up volume
volume up
turn down volume
volume down
mute
(/set) volume (0/off/mute)
(/set) volume (N)
(/set) volume (N) percent
unmute
(/set) volume (on/unmute)

...right?  Am I missing anything?

**Comment by michaelhinchey** [comment]
I would love this to work with my av receiver with harmony hub.

**Comment by digiltd** [comment]
Not sure, [this](https://github.com/alexa/skill-sample-nodejs-audio-player) might help. Wrong language but similar concepts.

There is always the option of a separate skill. I am looking at one that works like AnyMote. So you would say something like "alexa kodi remote mode" and then it responds to "up" "down" , "down" , "down" etc without the initial intent

**Comment by jingai** [comment]
There seems to be continuing confusion with what I'm trying to achieve here.. this Issue, and the feature it will become, is specifically for adjusting _Kodi's_ volume, not the connected A/V equipment.  For the latter, yes, some other skill would be used.

https://github.com/alexa/skill-sample-nodejs-audio-player doesn't appear to have any intents related to volume control at all.  Am I missing something?

Basically, my question was, because I cannot find a comprehensive list of the ways you can ask Alexa to adjust volume.. is the list I came up with everything we should support?

**Comment by jingai** [comment]
> There is always the option of a separate skill. I am looking at one that works like AnyMote. So you would say something like "alexa kodi remote mode" and then it responds to "up" "down" , "down" , "down" etc without the initial intent

I assume you mean make a separate skill using the Smart Home API?  I intend to play around with the audio player interface soon, which should also achieve this, but.. this is not within the scope of this particular Issue.

**Comment by digiltd** [comment]
So invoke this Kodi skill and send a command i.e. "alexa, tell kodi to set the volume to 5"?

**Comment by digiltd** [comment]
> I assume you mean make a separate skill using the Smart Home API? I intend to play around with the audio player interface soon, which should also achieve this, but.. this is not within the scope of this particular Issue.

No, something similar to this : https://www.youtube.com/watch?v=n2P9EkRGjfE

(though not as tedious)

**Comment by jingai** [comment]
This is all way off-topic for this Issue... let's please keep this on-topic.  If you want to continue discussing how things are done in that video (it's just keeping the session open), posting to the forum would be the right place.. but see [my post here](http://forum.kodi.tv/showthread.php?tid=254502&pid=2449734#pid2449734) on why I don't think we should do it this way.

...all I need to know for this Issue is whether the list of utterances I've come up with is comprehensive.

**Comment by digiltd** [comment]
agreed about the tangent

re the suggested utterances, i am not sure you are going to be able to use those types of commands. I don't know for sure, but I would expect them to be reserved and generally problematic.

Suck it and see i suppose :)

**Comment by jingai** [comment]
They seem to be working so far.  I just wanted to make sure I covered all of the ways people generally request volume changes with Alexa.  I suppose we'll find out soon on the forums if I missed anything :P

**Comment by jingai** [comment]
https://github.com/jingai/kodi-alexa/tree/setvol implements this.  I'd appreciate it if others could test.  I'll hold off on the merge for a day or two.
